File: mirror/tier2-python/showcase/sec_input.py
"""
MIRROR V3: Tier 2 Python Showcase
Category: Security
Module: sec_input.py
Purpose: Demonstrate input validation patterns
"""

from typing import Union, Any


def validate_input(value: str, rules: dict) -> bool:
    """Validate input against rules"""
    if not isinstance(value, str):
        return False
    
    if rules.get("min_length") and len(value) < rules["min_length"]:
        return False
    
    if rules.get("max_length") and len(value) > rules["max_length"]:
        return False
    
    if rules.get("not_empty") and len(value) == 0:
        return False
    
    if rules.get("alphanumeric") and not value.isalnum():
        return False
    
    return True


def sanitize_input(value: str) -> str:
    """Sanitize input string"""
    if not isinstance(value, str):
        return ""
    
    # Remove dangerous characters
    dangerous = ['<', '>', '"', "'", '&', ';']
    sanitized = value
    for char in dangerous:
        sanitized = sanitized.replace(char, '')
    
    return sanitized.strip()


def validate_email(email: str) -> bool:
    """Validate email format"""
    if not isinstance(email, str):
        return False
    if '@' not in email or '.' not in email:
        return False
    return True


result = {
    "valid_hello": validate_input("hello", {"min_length": 3, "max_length": 10}),
    "invalid_short": validate_input("hi", {"min_length": 3}),
    "invalid_empty": validate_input("", {"not_empty": True}),
    "sanitized": sanitize_input("hello<script>"),
    "valid_email": validate_email("user@example.com"),
    "input_validation_active": True,
}

if __name__ == "__main__":
    print(result)

File: mirror/tier2-ruby/showcase/sec_input.rb
# MIRROR V3: Tier 2 Ruby Showcase
# Category: Security
# Module: sec_input.rb
# Purpose: Demonstrate input validation patterns

def validate_input(value, rules)
  return false unless value.is_a?(String)
  
  return false if rules[:min_length] && value.length < rules[:min_length]
  return false if rules[:max_length] && value.length > rules[:max_length]
  return false if rules[:not_empty] && value.empty?
  return false if rules[:alphanumeric] && !value.match?(/\A[a-zA-Z0-9]*\z/)
  
  true
end

def sanitize_input(value)
  return "" unless value.is_a?(String)
  
  dangerous = ['<', '>', '"', "'", '&', ';']
  sanitized = value
  dangerous.each { |char| sanitized = sanitized.gsub(char, '') }
  
  sanitized.strip
end

def validate_email(email)
  return false unless email.is_a?(String)
  email.include?('@') && email.include?('.')
end

result = {
  valid_hello: validate_input("hello", {min_length: 3, max_length: 10}),
  invalid_short: validate_input("hi", {min_length: 3}),
  invalid_empty: validate_input("", {not_empty: true}),
  sanitized: sanitize_input("hello<script>"),
  valid_email: validate_email("user@example.com"),
  input_validation_active: true,
}

puts result.inspect
